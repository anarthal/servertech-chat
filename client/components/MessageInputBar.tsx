import { useCallback, useEffect, useRef } from "react";

// The input bar where the user types messages
export default function MessageInputBar({
  onMessage,
}: {
  onMessage: (msg: string) => void;
}) {
  const inputRef = useRef<HTMLInputElement>(null);

  const onKeyDown = useCallback(
    (event: KeyboardEvent) => {
      // Focus the input when the user presses any key, to make interaction smoother
      inputRef.current.focus();

      // Invoke the passed callback when Entr is hit, then clear the message
      if (event.key === "Enter") {
        const messageContent = inputRef.current?.value;
        if (messageContent) {
          onMessage(messageContent);
          inputRef.current.value = "";
        }
      }
    },
    [onMessage],
  );

  useEffect(() => {
    window.addEventListener("keydown", onKeyDown);
    return () => {
      window.removeEventListener("keydown", onKeyDown);
    };
  }, [onKeyDown]);

  return (
    <div className="flex">
      <div
        className="flex-1 flex p-2"
        style={{ backgroundColor: "var(--boost-light-grey)" }}
      >
        <input
          type="text"
          className="flex-1 text-xl pl-4 pr-4 pt-2 pb-2 border-0 rounded-xl bg-white"
          placeholder="Type a message..."
          ref={inputRef}
        />
      </div>
    </div>
  );
}
