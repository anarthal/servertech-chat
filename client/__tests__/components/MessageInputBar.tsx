import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import MessageInputBar from "@/components/MessageInputBar";
import "@testing-library/jest-dom";

describe("MessageInputBar", () => {
  test("renders message input bar", () => {
    // Render
    const { asFragment } = render(<MessageInputBar onMessage={jest.fn()} />);

    // The bar is there
    expect(
      screen.getByPlaceholderText("Type a message..."),
    ).toBeInTheDocument();

    // The send icon is there
    expect(screen.getByTestId("SendIcon")).toBeInTheDocument();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("keyboard interactions", async () => {
    const onMessage = jest.fn();
    const user = userEvent.setup();

    // Render
    render(<MessageInputBar onMessage={onMessage} />);

    // Typing autofocus the inputs
    await user.keyboard("this is a");

    // No callback until enter is hit
    expect(onMessage).toHaveBeenCalledTimes(0);

    // Hitting enter calls the callback
    await user.keyboard(" message{Enter}");
    expect(onMessage).toHaveBeenCalledWith("this is a message");
    expect(onMessage).toHaveBeenCalledTimes(1);

    // Hitting Enter cleared the input bar, so we can send another message
    onMessage.mockReset();
    await user.keyboard("another message!{Enter}");
    expect(onMessage).toHaveBeenCalledWith("another message!");
    expect(onMessage).toHaveBeenCalledTimes(1);
  });

  test("send icon interactions", async () => {
    const onMessage = jest.fn();
    const user = userEvent.setup();

    // Render
    render(<MessageInputBar onMessage={onMessage} />);

    // Type a message
    await user.keyboard("this is a message");

    // Clicking the send icon calls the callback
    screen.getByTestId("SendIcon").parentElement.click();
    expect(onMessage).toHaveBeenCalledWith("this is a message");
    expect(onMessage).toHaveBeenCalledTimes(1);

    // Sending the message cleared the input. We can now send another message
    onMessage.mockReset();
    await user.keyboard("another message!");
    screen.getByTestId("SendIcon").parentElement.click();
    expect(onMessage).toHaveBeenCalledWith("another message!");
    expect(onMessage).toHaveBeenCalledTimes(1);
  });
});
