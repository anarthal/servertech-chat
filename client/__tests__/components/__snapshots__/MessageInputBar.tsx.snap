// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInputBar renders message input bar 1`] = `
<DocumentFragment>
  <div
    class="flex"
  >
    <div
      class="flex-1 flex p-2"
    >
      <input
        class="flex-1 text-xl pl-4 pr-4 pt-2 pb-2 border-0 rounded-xl bg-white"
        placeholder="Type a message..."
        type="text"
      />
    </div>
  </div>
</DocumentFragment>
`;
